tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim[i],
+		}
+	}
+
+	initVal, err := gr.Op("Const", name+"/initial_value", nil, "", map[string]interface{}{
+		"dtype": ts.DataType(),
+		"value": ts,
+		"shape": shape,
+	})
+	if err != nil {
+		return
+	}
+
+	variable, err := gr.Op("Variable", name, nil, "", map[string]interface{}{
+		"dtype":       ts.DataType(),
+		"shape":       shape,
+		"container":   "",
+		"shared_name": "",
+	})
+	if err != nil {
+		return
+	}
+
+	variable.ref = variable.def
+
+	_, err = gr.Op("Assign", name+"/Assign", []*GraphNode{variable, initVal}, "", map[string]interface{}{
+		"use_locking":    true,
+		"validate_shape": true,
+	})
+	if err != nil {
+		return
+	}
+
+	op, err = gr.Op("Identity", name+"/read", []*GraphNode{variable}, "", nil)
+
+	// For reference this variable, use the variable as it.
+	op.ref = variable.def
+
+	return
+}
+
+// String Returns a string representation of this graph, used for debugging
+// proposals.
+func (gr *Graph) String() string {
+	var bufStr bytes.Buffer
+	proto.MarshalText(&bufStr, gr.def)
+
+	return bufStr.String()
+}
+
+// addInitializationGraphOp Add the initialization operation to the graph to
+// cover all the added variables
+func (gr *Graph) addInitializationGraphOp() {
+	inputs := make([]string, len(gr.variables))
+	i := 0
+	for input := range gr.variables {
+		inputs[i] = "^" + input + "/Assign"
+		i++
+	}
+
+	gr.def.Node = append(gr.def.Node, &pb.NodeDef{
+		Name:  "init",
+		Op:    "NoOp",
+		Input: inputs,
+	})
+}
+
+// Placeholder Adds a placeholder to the Graph, a placeholder is an
+// operation that must be fed with data on execution.
+func (gr *Graph) Placeholder(name string, dataType DataType, dims []int64, dimNames []string) (op *GraphNode) {
+	op = &GraphNode{
+		outDataTypes: map[string]DataType{
+			name: dataType,
+		},
+		def: &pb.NodeDef{
+			Name: name,
+			Op:   "Placeholder",
+			Attr: make(map[string]*pb.AttrValue),
+		},
+	}
+	op.def.Attr["dtype"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Type{
+			Type: pb.DataType(dataType),
+		},
+	}
+
+	shape := &pb.TensorShapeProto{
+		Dim: make([]*pb.TensorShapeProto_Dim, len(dims)),
+	}
+
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim,
+		}
+
+		if len(dimNames) == len(dims) {
+			shape.Dim[i].Name = dimNames[i]
+		}
+	}
+
+	op.def.Attr["shape"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Shape{
+			Shape: shape,
+		},
+	}
+
+	gr.def.Node = append(gr.def.Node, op.def)
+
+	return
+}
+
+// AsStr Returns the current graph serialized so it can be exported.
+func (gr *Graph) AsStr() []byte {
+	result, _ := proto.Marshal(gr.def)
+
+	return result
+}
+
+// ErrExpectedVarAsinput The specified operation is not defined.
+type ErrExpectedVarAsinput struct {
+	operation string
+	inputPos  int
+}
+
+func (e *ErrExpectedVarAsinput) Error() string {
+	return fmt.Sprintf(
+		"The expected input value at pos %d for the operation '%s' must be of type Variable",
```

Comment:
remove word 'expected' - the error is about the supplied input value, not an expected one.